### PR TITLE
feat: Add Session JWT support

### PR DESCRIPTION
**Pull request by jeremy-stytch**
The Stytch API now returns and accepts `session_jwt` values as JWT alternatives to opaque
`session_token` values. Support in this client library comes in two forms:

1. All the relevant request and response types have been updated.
2. There's a new `client.sessions.authenticateJwt` method for parsing and verifying a JWT in a
   way that falls back to calling the Stytch API if the token can't be verified locally.
3. The client library handles caching the project's verification key set, but there's also
   `client.sessions.jwks` for completeness.

Stytch session JWTs are valid for the lifetime of the session they were created for. The
`max_token_age` option controls how old a JWT can get before it's required to be validated with
the Stytch API. This gives applications some flexibility in handling JWTs for sessions that have
been recently revoked.

This also introduces a new `ClientError` for throwing errors that don't come from Stytch API
requests. They come with error codes to help identify common or interesting errors.

### CODE REVIEW

File: dist/sessions.js [hunk 1]
```diff
@@ -43,6 +59,87 @@ class Sessions {
...
+
+  async authenticate_jwt(jwt, options) {
+    try {
+      const session = await this.authenticate_jwt_local(jwt, options);
+      return {
+        session,
+        session_jwt: jwt,
+        session_token: "",
+        // Pretend that this was a successful API request.
+        request_id: "",
```
Comment (chris-stytch) [inline dist/sessions.js:77]: Our API will return a `request_id` here, imo we shouldn't overwrite as an empty string.

File: lib/sessions.ts [hunk 2]
```diff
@@ -85,6 +109,98 @@ export class Sessions {
     });
   }
 
+  /** Parse a JWT and verify the signature, preferring local verification over remote.
+   *
+   * If maxTokenAge is set, remote verification will be forced if the JWT was issued at (based on
+   * the "iat" claim) more than maxTokenAge seconds ago.
+   */
+  async authenticate_jwt(
```
Comment (cal-stytch) [inline lib/sessions.ts:117]: Should we have a force no matter what method as well? Perhaps for a sensitive action you'd want to be 100% sure the session is still valid (user did something on another platform)

Comment (cal-stytch) [reply lib/sessions.ts:117]: I suppose you could go back to the opaque session token for that (assuming the developer decided to hold onto it)

File: lib/sessions.ts [hunk 3]
```diff
@@ -85,6 +109,98 @@ export class Sessions {
...
+    }
+  ): Promise<AuthenticateResponse> {
+    try {
+      const session = await this.authenticate_jwt_local(jwt, options);
+      return {
+        session,
+        session_jwt: jwt,
+        session_token: "",
+
+        // Pretend that this was a successful API request.
```
Comment (Max-Stytch) [inline lib/sessions.ts:130]: Don't pretend- return a new/diff type! Feel free to only return the `session` object from this function call instead.

File: lib/sessions.ts [hunk 4]
```diff
@@ -85,6 +109,98 @@ export class Sessions {
...
+        request_id: "",
```
Comment (chris-stytch) [inline lib/sessions.ts:131]: `request_id` here as well.

File: lib/sessions.ts [hunk 5]
```diff
@@ -85,6 +109,98 @@ export class Sessions {
...
+        status_code: 200,
+      };
+    } catch (err) {
+      if (err instanceof ClientError && err.code === "jwt_too_old") {
+        // Token was too old (stale) to verify locally. Check with the Stytch API.
```
Comment (chris-stytch) [inline lib/sessions.ts:136]:
```suggestion
        // JWT was too old (stale) to verify locally. Check with the Stytch API.
```
Very much a nit, but prefer JWT here to "token" since our API has endpoints that accept a "token" value and that is not a JWT.

File: lib/sessions.ts [hunk 6]
```diff
@@ -85,6 +109,98 @@ export class Sessions {
...
+      if (now - iat > maxTokenAge) {
+        throw new ClientError(
+          "jwt_too_old",
+          `JWT was issued at ${iat}, more than ${maxTokenAge} seconds ago`
+        );
+      }
+    }
+
+    // Re-pack the JWT contents as session data. Since we're actually changing the types of
+    // values stored on the session object, we have to disable type-checking for a bit.
```
Comment (Max-Stytch) [inline lib/sessions.ts:182]: create a different object to return- Don't mutate this one in place

File: lib/sessions.ts [hunk 7]
```diff
@@ -85,6 +109,98 @@ export class Sessions {
...
+    //
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    const session = payload[sessionClaim] as any;
+
+    // The subject claim is the user ID.
+    session.user_id = payload.sub;
```
Comment (Max-Stytch) [inline lib/sessions.ts:188]: and if you are set on mutating in place, clear up the `sub` here like you clear up the `id` below

File: lib/sessions.ts (hunk 2 above)
Comment (cal-stytch) [inline lib/sessions.ts:117]: I believe the rest of the library does camelCase, so I believe this should be authenticateJwt (see loginOrCreate)

File: lib/sessions.ts [hunk 8]
```diff
@@ -85,6 +109,98 @@ export class Sessions {
...
+      throw err;
+    }
+  }
+
+  /** Parse a JWT and verify the signature locally (without making an /authenticate call).
+   *
+   * If maxTokenAge is set, this will return an error if the JWT was issued (based on the "iat"
+   * claim) more than maxTokenAge seconds ago.
+   */
+  async authenticate_jwt_local(
```
Comment (cal-stytch) [inline lib/sessions.ts:149]: See above, I believe this should be authenticateJwtLocal

File: test/manual.test.ts [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+      project_id: env("PROJECT_ID"),
+      secret: env("SECRET"),
+      env: process.env["STYTCH_API_URL"] || stytch.envs.test,
+    });
+  });
+
+  // To run these tests, start a new session (it doesn't matter how) and paste its session_token
+  // and session_jwt values here.
+  //
+  // Do not commit any real tokens to this repo.
```
Comment (Max-Stytch) [inline test/manual.test.ts:35]: Use `jose` to create a test JWKS and test JWTs to verify against inside the test itself. See [this file](https://github.com/stytchauth/web-backend/blob/3a0c3a2c07d436bf362e6b11556323dab20ee4fb/app/auth/utils/jwt.test.ts) for an example

File: dist/sessions.js [hunk 10]
```diff
@@ -43,6 +59,87 @@ class Sessions {
...
+      const iat = payload.iat;
+
+      if (!iat) {
+        throw new _errors.ClientError("jwt_invalid", "JWT was missing iat claim");
+      }
+
+      if (now - iat > maxTokenAge) {
+        throw new _errors.ClientError("jwt_too_old", `JWT was issued at ${iat}, more than ${maxTokenAge} seconds ago`);
+      }
+    } // Re-pack the JWT contents as session data. Since we're actually changing the types of
```
Comment (chris-stytch) [inline dist/sessions.js:123]: New line for this comment here.

File: test/manual.test.ts (hunk 9 above)
Comment (Max-Stytch) [reply test/manual.test.ts:35]: We should absolutely have non-manual tests for `authenticate_jwt_local` which doesn't need to make any network calls

**Review by Max-Stytch — CHANGES_REQUESTED** [review]

File: lib/sessions.ts (hunk 2 above)
Comment (jeremy-stytch) [reply lib/sessions.ts:117]: > Should we have a "force remote verification no matter what" method as well? Perhaps for a sensitive action you'd want to be 100% sure the session is still valid (user did something on another platform)

We do! In this method, you can set `maxTokenAge: 0`. Or you can call the `authenticate` method (always uses the network) with `{session_jwt: jwt}`. 

I'll update the doc comment here to add that usage note, but I'm happy to keep discussing usability for that flow.

File: lib/sessions.ts (hunk 3 above)
Comment (jeremy-stytch) [reply lib/sessions.ts:130]: Oh good point! I guess we don't actually need compatibility with `authenticate` on this one. I'll do `{session, session_jwt}` here, since I think it'll be useful to have the new JWT if we do go to the network.

File: lib/sessions.ts (hunk 5 above)
Comment (jeremy-stytch) [reply lib/sessions.ts:136]: ... I have no explanation for why I said "token" here. Thanks for catching it!

File: lib/sessions.ts (hunk 4 above)
Comment (jeremy-stytch) [reply lib/sessions.ts:131]: This one is going away (see Max's thread above), but this was part of the "pretend that this was a successful API request". If the JWT validated locally, we wouldn't have gone to the API at all, so there wouldn't have been an API request ID to overwrite.

File: test/manual.test.ts (hunk 9 above)
Comment (jeremy-stytch) [reply test/manual.test.ts:35]: Thanks for the link! I had debated exposing `currentDate` to make this testable, but actually minting JWTs is an even better test.

**Review by jeremy-stytch — COMMENTED** [review]
Thanks for the reviews! I'll ping or resolve the threads once I've handled them. I'll be pushing multiple times, so feel free to ignore this until you get a real re-review request from me.

File: dist/sessions.js (hunk 1 above)
Comment (jeremy-stytch) [reply dist/sessions.js:77]: Whoops, I missed this when I was going through the diff view because GH auto-collapses compiled code (`dist` and `types`) set by our `.gitattributes` file. I think this is the generated equivalent of [this](https://github.com/stytchauth/stytch-node/pull/116#discussion_r831672603).

File: dist/sessions.js (hunk 10 above)
Comment (jeremy-stytch) [reply dist/sessions.js:123]: This one's in the compiled output, so I'm going to skip it. If you feel strongly about it, my browser history tells me I've looked into this before lol: https://github.com/babel/babel/issues/8437#issuecomment-623853285

Comment (jeremy-stytch) [reply dist/sessions.js]: The whole comment went away, so I get to resolve this for free 😛

**Review by chris-stytch — APPROVED** [review]
All of my comments have been resolved.

File: test/manual.test.ts (hunk 9 above)
Comment (jeremy-stytch) [reply test/manual.test.ts:35]: I ended up doing both. I think that exposing `currentDate` (as `current_date`) only on `authenticateJwtLocal` is the right balance of control and convenience. `authenticateJwt` (auto-fallback) always uses the current time, but an app may want to a frozen local time and fall back to a network call separately.

File: lib/sessions.ts [hunk 11]
```diff
@@ -85,6 +145,101 @@ export class Sessions {
...
+      current_date?: Date;
+    }
+  ): Promise<Session> {
+    const now = options?.current_date || new Date();
+
+    let payload;
+    try {
+      const token = await jose.jwtVerify(jwt, this.jwksClient, {
+        ...this.jwtOptions,
+        currentDate: now,
```
Comment (Max-Stytch) [inline lib/sessions.ts:202]: I think we might want to expose `clockTolerance` here too - it won't have much of an effect with long-lived tokens, but certainly with short-lived tokens 30s of clock skew can go a long way

File: test/sessions.test.ts [hunk 12]
```diff
@@ -113,3 +134,159 @@ describe("sessions.revoke", () => {
...
+  let startedAt: Date;
+  let expiresAt: Date;
+
+  beforeEach(async () => {
+    // Generate a new key and add it to a local JWKS.
+    const keyID = "key0";
+    const { publicKey, privateKey } = await jose.generateKeyPair("RS256");
+    const jwk = (publicKey as crypto.KeyObject as any).export({
+      format: "jwk",
+    });
```
[the remark this answers is not in the document]
Comment (Max-Stytch) [reply test/sessions.test.ts:196]: Hm - when I run these tests locally I get 
```
  ● sessions.authenticateJwtLocal › zero max_token_age_seconds forces staleness

    TypeError [ERR_INVALID_OPT_VALUE]: The value "jwk" is invalid for option "format"

      192 |     const keyID = "key0";
      193 |     const { publicKey, privateKey } = await jose.generateKeyPair("RS256");
    > 194 |     const jwk = (publicKey as crypto.KeyObject as any).export({
          |                                                              ^
      195 |       format: "jwk",
      196 |     });
      197 |     const jwks = jose.createLocalJWKSet({ keys: [{ ...jwk, kid: keyID }] });

```

**Review by Max-Stytch — COMMENTED** [review]
One last thing that I see sometimes is making sure the JWK/JWT alg used is sufficiently strong. I would feel pretty good about adding tests showing/documenting that `none`-alg tokens always error out

File: lib/sessions.ts (hunk 11 above)
Comment (jeremy-stytch) [reply lib/sessions.ts:203]: Makes sense to me! Implemented in [9b43c6d](https://github.com/stytchauth/stytch-node/pull/116/commits/9b43c6d23a0d4708bd980436fdc69723ccc4e305)

File: test/sessions.test.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (jeremy-stytch) [reply test/sessions.test.ts:194]: Ah looks like this is a Node version difference because this got added in v15 (v16 for most production usage). Our tests run on v16. I'll add a comment for when we decide to run automated tests on older versions of Node.

For now, I've manually tested `authenticateJwt` in the latest release of each supported LTS:
- v12.22.11
- v14.19.1
- v16.14.2

**Review by Max-Stytch — APPROVED** [review]
